common/src/main/java/com/scottlogic/deg/generator/fieldspecs/whitelist/FrequencyDistributedSet.java

```diff
@@ -0,0 +1,133 @@
+
+    private final List<WeightedElement<T>> underlyingCumulativeWeights;
+
+    public FrequencyDistributedSet(final Set<WeightedElement<T>> underlyingWeights) {
+        if (underlyingWeights.contains(null)) {
+            throw new IllegalArgumentException("FrequencyDistributedSet should not contain null elements");
+        }
+
+        Set<WeightedElement<T>> normalisedWeights = normalise(underlyingWeights);
+        this.underlyingWeights = Collections.unmodifiableSet(normalisedWeights);
+        this.underlyingCumulativeWeights = cumulative(normalisedWeights);
+    }
+
+    private static <T> Set<WeightedElement<T>> normalise(final Set<WeightedElement<T>> denormalised) {
+        final double total = denormalised.stream()
+            .map(WeightedElement::weight)
+            .reduce(0.0D, Double::sum);
+
+        // Stream (even with 0 elements) ensure a copy of the original set is returned
+        return denormalised.stream()
+            .map(holder -> new WeightedElement<>(holder.element(), holder.weight() / total))
+            .collect(Collectors.toSet());
+    }
+
+    public static <T> FrequencyDistributedSet<T> singleton(final T element) {
+        return FrequencyDistributedSet.uniform(Collections.singleton(element));
+    }
+
+    public static <T> FrequencyDistributedSet<T> uniform(final Collection<T> underlyingSet) {
+        return new FrequencyDistributedSet<>(
+            underlyingSet.stream()
+                .map(WeightedElement::withDefaultWeight)
+                .collect(Collectors.toSet()));
+    }
+
+    private static <T> List<WeightedElement<T>> cumulative(Set<WeightedElement<T>> nonCumulative) {
+        List<WeightedElement<T>> cumulative = new LinkedList<>();
+        double runningTotal = 0.0D;
+        for (WeightedElement<T> holder : nonCumulative) {
+            runningTotal += holder.weight();
+            cumulative.add(new WeightedElement<>(holder.element(), runningTotal));
+        }
+
+        if (!cumulative.isEmpty()) {
+           replaceLastCumulativeElement(cumulative);
+        }
+
+        return new ArrayList<>(cumulative);
+    }
+
+    private static <T> void replaceLastCumulativeElement(List<WeightedElement<T>> cumulative) {
+        int lastIndex = cumulative.size() - 1;
+        WeightedElement<T> last = cumulative.get(lastIndex);
+        cumulative.remove(lastIndex);
+        cumulative.add(new WeightedElement<>(last.element(), 1.0D));
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T> FrequencyDistributedSet<T> empty() {
+        return (FrequencyDistributedSet<T>) EMPTY;
+    }
+
+    @Override
+    public Set<WeightedElement<T>> distributedSet() {
+        return underlyingWeights;
+    }
+
+    @Override
+    public T pick(double random) {
+        final int index = binarySearchForBucket(1.0D - random);
+        return underlyingCumulativeWeights.get(index).element();
+    }
+
+    private int binarySearchForBucket(final double value) {
```

Comment:
what is a bucket

Reply:
[(https://en.wikipedia.org/wiki/Bucket_(computing))|Bucket]